src/ray/common/asio/io_context.h

```diff
@@ -0,0 +1,108 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <boost/asio.hpp>
+#include <limits>
+#include "absl/container/flat_hash_map.h"
+#include "absl/synchronization/mutex.h"
+#include "ray/common/ray_config.h"
+#include "ray/util/logging.h"
+
+/// Count, queueing, and execution statistics for an asio handler.
+struct HandlerStats {
+  // Counts.
+  int64_t cum_count = 0;
+  int64_t curr_count = 0;
+
+  // Queue stats.
+  int64_t cum_queue_time = 0;
+  int64_t min_queue_time = std::numeric_limits<int64_t>::max();
+  int64_t max_queue_time = -1;
+
+  // Execution stats.
+  int64_t cum_execution_time = 0;
+  int64_t min_execution_time = std::numeric_limits<int64_t>::max();
+  int64_t max_execution_time = -1;
+};
+
+/// A mutex wrapper around a handler stats struct.
+struct GuardedHandlerStats {
+  // Stats for some handler.
+  HandlerStats stats;
+
+  // The mutex protecting the reading and writing of these stats.
+  // This mutex should be acquired with a reader lock before reading, and should be
+  // acquired with a writer lock before writing.
+  mutable absl::Mutex mutex;
+};
+
+/// A proxy for boost::asio::io_context that collects statistics about posted handlers.
+class io_context_proxy : public boost::asio::io_context {
```

Comment:
Can you rename this file to instrumented_io_context? And also the class accordingly? Proxy is less specific of a name.